### PR TITLE
chore: migrate from master branch to main

### LINKED ISSUE #8153: Proposal: migrate renovatebot/renovate from master to main branch
**Proposal: migrate `renovatebot/renovate` from `master` branch to `main` branch:**

GitHub is preparing to migrate existing repositories from `master` to `main`.
I think it's a good idea to prepare/think about a migration from `renovatebot/renovate master` to `renovatebot/renovate main`.

The Renovate bot itself already knows how to deal with a `main` branch, so this is not about the Renovate bot itself, only about the GitHub repository `renovatebot/renovate`.

**What GitHub will do for us (starting January 2021):**

- GitHub will redirect any old links that target `master` to point to `main`
- GitHub Pages can build and deploy from any branch
- Migrate open PRs to target `main` instead of `master`
- Migrate branch protections rules that explicitly target `master`
- Contributors will get a prompt helping them to update their local Git configuration

**What we must do ourselves:**

Even if we use the GitHub automation, there's still some things that will need to be migrated by hand:

- [ ] ci: target `main` instead of `master` (GitHub actions)
- [ ] ci: Update build and release scripts (semantic release bot)
- [ ] chore: Ensure other Renovate projects use `renovate/renovate main` instead of `renovate/renovate master`
- [ ] docs: Update references to Renovate `master` branch in documentation

**Did you already have any implementation ideas?**

See checklist above for my initial thoughts on the scope of this work.

I was going to do a lot of manual work to update our references to `master` to `main` in CI scripts, build scripts and documentation, but I think it's smarter to discuss things first before starting work... 😄 

**Documentation**

- GitHub has official documentation explaining what GitHub will do automatically for us at [github/renaming](https://github.com/github/renaming).
- The ProGit2 book has a chapter called "Changing the master branch name" that explains the pure Git side of things https://git-scm.com/book/en/v2/Git-Branching-Branch-Management

**Discussion**

If you don't want to migrate the `renovatebot/renovate` repository from using a `master` branch to using a `main` branch, feel free to close this issue.
This is just me "thinking aloud" on what it would take to migrate.

### PR DESCRIPTION
**Pull request by HonkingGoose**
## Changes:

- `master` -> `main` for CI
- Replace links to `renovatebot/renovate master` to point to `renovatebot/renovate main` instead

Really rough draft to get an idea of the changes needed to migrate from `master` to `main` on this repo.

Maybe @rarkins can make a complete copy of this repo somewhere safe and try this out, and see where stuff breaks.
I have **not** tested this myself yet, as I would need to replicate the full setup that this repository has somehow (external bots, branch protection rules, semantic release bot, etc.)

I have also **not replaced** any mentions of `master` in the `docs/usage` section or in any fixtures or snapshots, as that pertains to the code of Renovate bot itself and what branches the bot supports on specific platforms.

## Context:

Explore what changes we would probably need to migrate from `master` to `main` as the default branch on the `renovatebot/renovate` repository.

Closes #8153.

Also for more information read: https://github.com/github/renaming
GitHub docs also has some guidance on how to rename a branch from the interface as admin: https://docs.github.com/en/github/administering-a-repository/renaming-a-branch

## Checklist:

GitHub should do the following for us automatically:

- GitHub will redirect any old links that target `master` to point to `main`
- GitHub Pages can build and deploy from any branch
- Migrate open PRs to target `main` instead of `master`
- Migrate branch protections rules that explicitly target `master`
- Contributors will get a prompt helping them to update their local Git configuration

What we need to do before migrating:

- [ ] Check if there's other external config for bots in use on this repo (bot dashboards, code-coverage bot etc.)
- [ ] Check if all the relevant CI files are correct for use with `main`

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added unit tests, or
- [ ] No new tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by viceice** [comment]
needs branch protection added for main, should be doable before exists

**Comment by HonkingGoose** [comment]
Yeah the branch protection rules for `main` should be setup **before** merging.

**Review by viceice — DISMISSED** [review]
Assuming you used search and replace ;)

**Comment by HonkingGoose** [comment]
I used the search funtion in VSCode, with the following settings:

Search for string `master`.
Files to exclude: `__fixtures__, .spec.ts, __snapshots__`

**Comment by HonkingGoose** [comment]
GitHub will migrate the branch protection rules automatically if you're doing the migration from `master` to `main` via the GitHub interface (see image): https://github.com/github/renaming/blob/main/README.md#renaming-existing-branches

**Review by JamieMagee — DISMISSED** [review]

**Comment by HonkingGoose** [comment]
Merged the latest `master` branch and resolved the merge conflict, that was caused by the newlines change in the label bot file: `label-actions.yml`. I've kept the newlines from `master` as having 2 newlines makes it so that the bot uses paragraphs.

**Comment by JamieMagee** [comment]
@HonkingGoose This looks ready for merge, right? But it should be manually merged by @rarkins, as there are manual steps to follow.

**Comment by HonkingGoose** [comment]
@JamieMagee Yeah I think this is ready for final review and merge.
Do you have any tips on the migration @JamieMagee?

The main idea, I think, would be to freeze everything, turn off all the bots, and migrate via GitHub interface, merge this PR, confirm tests work, and then enable bots one by one, with the Semantic Release bot as the last one to be re-enabled.

Proposed steps:

1. Stop manually merging things into `master`, (basically a full freeze).
1. Turn off all bots that do things on our `master` branch (Renovate bot, Semantic Release, other CI).
1. Use the GitHub interface to migrate from `master` to `main`.
1. Confirm branch protections are still good on `main`.
1. GitHub will re-target all open pull requests (including this one).
1. Merge this PR into `main` so that all the CI uses the new `main` branch.
1. Confirm branch status checks work properly again.
1. Turn on bots one by one (do Semantic Release bot last).

**Comment by JamieMagee** [comment]
That sounds like a comprehensive migration plan. Could I suggest one more thing: can we try this on one of the other repos first as a test run for the main renovate repository?

**Comment by rarkins** [comment]
Not sure semantic release needs to be turned off as long as we don't add any release-triggering commits to master/main?

And yes maybe let's do all the smaller repos first and finish with this one.

**Comment by HonkingGoose** [comment]
> Could I suggest one more thing: can we try this on one of the other repos first as a test run for the main renovate repository?

Good idea! It's probably safer to try the smaller repos first, see if we've got all things covered, and as the last step migrate the main renovate repository.

> Not sure semantic release needs to be turned off as long as we don't add any release-triggering commits to master/main?

I'm trying to make sure that we don't create releases unnecessarily. But if we don't use any `fix:` or `feat:` or `refactor:` commits we should probably be safe?

**Comment by rarkins** [comment]
I updated the renovatebot/pep440 repo to use `main` and it went smoothly. e.g. PRs in progress were retargeted, actions continued fine.

This guide for local development pops up:

image

**Comment by JamieMagee** [comment]
Time to do the other repos, followed by this one?

**Comment by HonkingGoose** [comment]
Merged in current `master` and checked if the new `stale-action` needs anything (it doesn't as it doesn't rely on the concept of a branch😄).

**Comment by rarkins** [comment]
I think I've renamed all our repos except this one plus these:
- https://github.com/renovatebot/renovatebot.github.io
- https://github.com/renovatebot/docker-renovate-full
- https://github.com/renovatebot/docker-buildpack
- https://github.com/renovatebot/internal-tools
- https://github.com/renovatebot/.github

@viceice do you think they're all safe for a simple rename?

**Comment by viceice** [comment]
> I think I've renamed all our repos except this one plus these:
> - https://github.com/renovatebot/renovatebot.github.io

Needs workflow update

> - https://github.com/renovatebot/docker-renovate-full

This too

> - https://github.com/renovatebot/docker-buildpack 

This too

> - https://github.com/renovatebot/internal-tools 

This too

> - https://github.com/renovatebot/.github

 Should be safe.

> 
> @viceice do you think they're all safe for a simple rename?

**Comment by rarkins** [comment]
@viceice I've been checking them all out and updating workflows so no worries there. I was just concerned in case any of these are referred to by *other* repositories in workflows and could break if they somehow hardcode master

**Comment by viceice** [comment]
I'm pretty sure they don't hardcode any master reference beside the workflow

**Comment by HonkingGoose** [comment]
GitHub _should_ also redirect links pointing to old `master` to new `main` automatically.

Quote from https://github.com/github/renaming#whats-already-changed

> We've updated GitHub.com to redirect links that contain the deleted branch name such as master to the corresponding link in the repository's new default branch.

**Comment by HonkingGoose** [comment]
We should probably put up a pinned issue on the repositories we've just changed/are about to change.

I'm thinking of something like this:

```markdown
## Help the branch was renamed how do I fix my fork?

It's easy to fix your fork, as you can do most of the work straight from the GitHub web interface.

1. Go to your forks "homepage" on GitHub.
2. GitHub will have a popup over your `master` branch name, to tell you that the upstream repository has changed its default branch name.
3. Rename your forks branch from `master` to `main` from within the GitHub interface (Settings -> Branches -> Default branch -> click on the pencil icon to rename.
4. Let GitHub automatically rename your fork's branch name.
5. Go back to your fork's homepage, and follow the instructions GitHub gives you to rename the branch name on your local clone of your fork.
```

**Comment by viceice** [comment]
manual edit of `.\.git\refs\remotes\origin\HEAD` and change master to main. git won't do this automatically 😕

**Comment by rarkins** [comment]
GitHub provides a nice 3-line command to run locally to switch an existing clone from master to main - I'll post it here

**Comment by viceice** [comment]
> GitHub provides a nice 3-line command to run locally to switch an existing clone from master to main - I'll post it here

Yes, but it won't change the mentioned refs 😕

**Comment by rarkins** [comment]
> > GitHub provides a nice 3-line command to run locally to switch an existing clone from master to main - I'll post it here
> 
> Yes, but it won't change the mentioned refs 😕

Here's what it is, in case it helps anyway:

```sh
git branch -m master main
git fetch origin
git branch -u origin/main main
```

**Comment by viceice** [comment]
Found a solution: `git remote set-head origin main` 🎉 

https://gist.github.com/libitte/cbde168d26bc5faf9bf9fef648091b42

**Comment by JamieMagee** [comment]
👍 to @HonkingGoose's pinned issue idea. Also 👍 to including GitHub's information about switching your local clone, and @viceice's refs information.

Is 1 week's notice enough for the change in this repository?

**Comment by HonkingGoose** [comment]
> Is 1 week's notice enough for the change in this repository?

@JamieMagee 1 week's notice in a pinned issue _before_ starting is fine.

I'd say keep the issue pinned and open for a month _afterwards_, so that the frequent contributors can find the help quickly.

After a month we unpin and close the issue and link to it when somebody asks for help on how to fix their fork.

**Comment by viceice** [comment]
`.releaserc` needs `"branches": ["main"]`

**Comment by HonkingGoose** [comment]
> `.releaserc` needs `"branches": ["main"]`

You're right, good catch! 😄 

@viceice Can you check if the current `.releaserc` config is correct?

File: .releaserc [hunk 1]
```diff
@@ -1,4 +1,7 @@
 {
+  branches: [
+    {name: 'main'},
+  ],
```
Comment (viceice) [inline .releaserc:4]:
```suggestion
  "branches": ["main"],
```

This is enough

**Comment by viceice** [comment]
i think the v25 release will be a good point to switch to `main`

**Comment by HonkingGoose** [comment]
I've merged the latest changes from `master` into this branch as there were some changes related to making a build and release on the `v24` branch that I needed to pull in.

I've fixed the "bad merge" in the `.releaserc` file (it had two objects keys for the `branches` object...). We now have one `branches` object with configuration for the `v24` branch and the `main` branch.

This should be OK again, but please check my work.

**Review by viceice — CHANGES_REQUESTED** [review]
needs update, maybe deconflicting 😉

**Review by viceice — DISMISSED** [review]

**Review by viceice — APPROVED** [review]

**Comment by viceice** [comment]
new step popped up:
```bash
git branch -m master main
git fetch origin
git branch -u origin/main main
git remote set-head origin -a
```

**Comment by HonkingGoose** [comment]
Well that went smoothly! The build still works, the semantic release bot kept humming along, and things were fine. 😄 🎉 ✨ ✨ ✨ 

Thank you @viceice, @rarkins and @JamieMagee for the reviews, tips and suggestions to get this landed. ❤️

**Pull request merged**